### PR TITLE
[#155034260,#155034292,#153596281] Configures Kubernetes to publish the italia-backend app

### PR DESCRIPTION
**Pull request by cloudify**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -494,3 +495,46 @@ To deploy the documentation site (via GitHub pages):
...
+
+## Troubleshooting
+
+### Terraform
+
+#### Error locking state: Error acquiring the state lock
+
+It may happen that during a Terraform apply operation, something happens and the
+lock acquired by Terraform on the remote state blog doesn't get released.
+When this happen you'll get an error like this:
```
Comment (gunzip) [inline README.md]: happen(s) ?

File: architecture/decisions/0013-we-decide-how-to-expose-the-app-backend-to-the-internet.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+The `italia-app` application relies on a backend application ([italia-backend](https://github.com/teamdigitale/italia-backend))
+for intermediating the interaction with external services (e.g. the Digital
+Citizenship APIs) and for coordinating the SPID authentication process.
+
+## Decision
+
+The backend application (`app_backend`) is deployed as a Docker container
+inside a Kubernetes cluster, provisioned by Terraform.
+
+![img](https://www.planttext.com/plantuml/svg/XP7FIiGm4CRlynJp0cK5zo2oo887tSkoO0-2I4ZJsJPicyH9ilW7tzsaQbjwKCWXpFpovPlvjZv83h7l08oj2NMGdPP2EWTmPDaJolV899WQRGr-_60kLfrMGe_KALR4XW9veRhe0_78QjCmUIyyyLah6dMT4vLK9pArSBOUyLaTuFFtm6GCDqLHt4mMy1glBbRtPNdu6rglBmVg0M0gRpArSFWjMDuU_kUoPWPXsHhusIWpasdCGCYwQGFQ8pujdZu3xkzs-qTaEXFxj6kshs-0QQMzDZ9j68SfK5bZI8KK2o3Rc1jCBv5ym4fwgZ7brLeVLaw65hA7_370DbrJu5y0)
```
Comment (gunzip) [inline architecture/decisions/0013-we-decide-how-to-expose-the-app-backend-to-the-internet.md]: I'd rather download the SVG and put it in this repo

File: architecture/decisions/0013-we-decide-how-to-expose-the-app-backend-to-the-internet.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+[source of diagram](https://www.planttext.com/?text=XP7FIiGm4CRlynJp0cK5zo2oo887tSkoO0-2I4ZJsJPicyH9ilW7tzsaQbjwKCWXpFpovPlvjZv83h7l08oj2NMGdPP2EWTmPDaJolV899WQRGr-_60kLfrMGe_KALR4XW9veRhe0_78QjCmUIyyyLah6dMT4vLK9pArSBOUyLaTuFFtm6GCDqLHt4mMy1glBbRtPNdu6rglBmVg0M0gRpArSFWjMDuU_kUoPWPXsHhusIWpasdCGCYwQGFQ8pujdZu3xkzs-qTaEXFxj6kshs-0QQMzDZ9j68SfK5bZI8KK2o3Rc1jCBv5ym4fwgZ7brLeVLaw65hA7_370DbrJu5y0)
+
+Requests coming from the app to the backend gets routed through a few components:
```
Comment (gunzip) [inline architecture/decisions/0013-we-decide-how-to-expose-the-app-backend-to-the-internet.md]: get

File: architecture/decisions/0013-we-decide-how-to-expose-the-app-backend-to-the-internet.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  1. An Azure public IP with a firewall configured to listen on ports 80 and 443
+  1. A K8S service that routes the ports 80 and 443 to the Ingress
+  1. A K8S Ingress that terminates the HTTPS connection and routes the request based on the HTTP `Host` and path
```
Comment (gunzip) [inline architecture/decisions/0013-we-decide-how-to-expose-the-app-backend-to-the-internet.md]: link to https://kubernetes.io/docs/concepts/services-networking/ingress/ ?

File: architecture/decisions/0013-we-decide-how-to-expose-the-app-backend-to-the-internet.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+## Consequences
+
+### Responsibilities of the backend application
+
+  * it should listen on port 80 alone (since the Ingress terminates the HTTPS
+    connection)
+  * it should inspect `X-Forwarded-For` headers(s) in case it wants to know
+    something about the client
+  * avoid responding to non HTTPS requests
```
Comment (gunzip) [inline architecture/decisions/0013-we-decide-how-to-expose-the-app-backend-to-the-internet.md]: http ?

File: infrastructure/azure.tf [hunk 6]
```diff
@@ -850,3 +851,28 @@ module "kubernetes" {
   service_principal_client_id = "${data.azurerm_client_config.current.client_id}"
   service_principal_client_secret = "${var.ARM_CLIENT_SECRET}"
 }
+
+#
+# Allocates a public IP for exposing the Kubernetes services
+#
+# This IP needs to be registered on the following CNAMEs:
+#
+# *.k8s.test.cd.teamdigitale.it (for the test environment)
```
Comment (gunzip) [inline infrastructure/azure.tf]: is this going to change with *italia.it someday ?

File: infrastructure/kubernetes/README.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+## Additional components
+
+### Installing Cert Manager
+
+We use [Cert Manager](https://github.com/jetstack/cert-manager) for
+automatically provision the TLS certificates for public HTTPS endpoints.
+
+Cert Manager is [deployed](https://github.com/jetstack/cert-manager/blob/master/docs/user-guides/deploying.md)
+via [Helm](https://github.com/kubernetes/helm) with the
```
Comment (gunzip) [inline infrastructure/kubernetes/README.md]: I'll add avery short description of what Heml and Tiller are

File: infrastructure/kubernetes/ingress.yml [hunk 8]
```diff
@@ -161,12 +168,17 @@ kind: Service
...
   name: ingress-nginx-service
 spec:
+  # The LoadBalancer resource will be automatically created by Azure, allocating
+  # a public IP address, in case you need a specific IP address you should
+  # create the public IP adddress resource yourself and uncomment the
+  # loadBalancerIP attribute.
   type: LoadBalancer
-  # loadBalancerIP: TODO from static IP via Terraform
+  # loadBalancerIP = output of "azurerm_kubernetes_public_ip_ip" in "test" environment
+  loadBalancerIP: 13.95.82.27
```
Comment (gunzip) [inline infrastructure/kubernetes/ingress.yml]: is this fixed between runs ?

File: infrastructure/azure.tf (hunk 6 above)
Comment (cloudify) [reply infrastructure/azure.tf]: You mean the IP? it's up to us to decide, until we delete the Azure resource we can keep using that IP.

If we have a single Kubernetes cluster for all environments we could even use a single IP to route traffic to pods in different namespaces/environments.

File: infrastructure/kubernetes/ingress.yml (hunk 8 above)
Comment (cloudify) [reply infrastructure/kubernetes/ingress.yml]: yes it's a resource provisioned by Terraform (see `azure.tf`), so until the resource doesn't get deleted we can hold that IP

Comment (gunzip) [reply infrastructure/kubernetes/ingress.yml]: I'll add a line in the readme to check that this IP is in sync with the terraform resource and what to do to check it and eventually to keep them in sync

**Pull request merged**